proxy/src/transport/io.rs

```diff
@@ -0,0 +1,169 @@
+use std::io;
+use std::net::{Shutdown, SocketAddr};
+
+use bytes::Buf;
+use futures::Poll;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use super::AddrInfo;
+use self::internal::Io;
+
+/// A public wrapper around a `Box<Io>`.
+///
+/// This type ensures that the proper write_buf method is called,
+/// to allow vectored writes to occur.
+#[derive(Debug)]
+pub struct BoxedIo(Box<Io>);
+
+impl BoxedIo {
+    pub fn new<T: Io + 'static>(io: T) -> Self {
+        BoxedIo(Box::new(io))
+    }
+
+    /// Since `Io` isn't publicly exported, but `Connection` wants
+    /// this method, it's just an inherent method.
+    pub fn shutdown_write(&mut self) -> Result<(), io::Error> {
+        self.0.shutdown_write()
+    }
+}
+
+impl io::Read for BoxedIo {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+
+impl io::Write for BoxedIo {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.0.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.0.flush()
+    }
+}
+
+impl AsyncRead for BoxedIo {
+    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+        self.0.prepare_uninitialized_buffer(buf)
+    }
+}
+
+impl AsyncWrite for BoxedIo {
+    fn shutdown(&mut self) -> Poll<(), io::Error> {
+        self.0.shutdown()
+    }
+
+    fn write_buf<B: Buf>(&mut self, mut buf: &mut B) -> Poll<usize, io::Error> {
+        // A trait object of AsyncWrite would use the default write_buf,
+        // which doesn't allow vectored writes. Going through this method
+        // allows the trait object to call the specialized write_buf method.
+        self.0.write_buf_erased(&mut buf)
+    }
+}
+
+impl AddrInfo for BoxedIo {
+    fn local_addr(&self) -> Result<SocketAddr, io::Error> {
+        self.0.local_addr()
+    }
+
+    fn get_original_dst(&self) -> Option<SocketAddr> {
+        self.0.get_original_dst()
+    }
+}
+
+pub(super) mod internal {
+    use std::io;
+    use tokio::net::TcpStream;
+    use super::{AddrInfo, AsyncRead, AsyncWrite, Buf, Poll, Shutdown};
+
+    /// This trait is private, since it's purpose is for creating a dynamic
+    /// trait object, but doing so without care can lead not getting vectored
+    /// writes.
+    ///
+    /// Instead, used the concrete `BoxedIo` type.
+    pub trait Io: AddrInfo + AsyncRead + AsyncWrite + Send {
+        fn shutdown_write(&mut self) -> Result<(), io::Error>;
+
+        /// This method is to allow using `Async::write_buf` even through a
+        /// trait object.
+        fn write_buf_erased(&mut self, buf: &mut Buf) -> Poll<usize, io::Error>;
+    }
+
+    impl Io for TcpStream {
+        fn shutdown_write(&mut self) -> Result<(), io::Error> {
+            TcpStream::shutdown(self, Shutdown::Write)
+        }
+
+        fn write_buf_erased(&mut self, mut buf: &mut Buf) -> Poll<usize, io::Error> {
+            self.write_buf(&mut buf)
```

Comment:
Sorry I didn't comment on this before it was merged.

1. Is it possible to write this as `<Self as AsyncWrite + Sized>::write_buf(self, buf)`? IMO that would communicate precisely what we we're aiming for, if it works.

2. NIT: Does this need to be `&mut buf` here or is `buf` sufficient?

Reply:
👍 for the more explicit self type.

Reply:
It *does* need to be `&mut buf`, since the generic method signature is `write_buf<B: Buf>(&mut self, buf: &mut B)`, passing just `&mut Buf` maps it as `write_buf::<Buf>(buf)`. The requires a sized typed, so it must be called as `write_buf::<&mut Buf>(&mut buf)`. 😢